### ISSUE TITLE
Tendermint panics on TMSP App kill

### ISSUE THREAD

**Issue by srmo** [tendermint/tendermint#347]
Hi, when killing / shutting down our TMSP App (Java), tendermint 0.7.4 crashes with
```NOTE[12-29|10:54:21] Finalizing commit of block with 0 txs    module=consensus height=23 hash=A487C54C166B9352C1BCD2C5F106F0B175332B28
WARN[12-29|10:54:21] Error in proxyAppConn.EndBlock           module=state error=EOF
panic: Paniced questionably: Exec failed for application: EOF

goroutine 65 [running]:
panic(0x5691c0, 0xc421314d60)
    /usr/local/go/src/runtime/panic.go:500 +0x1a1
github.com/tendermint/tendermint/vendor/github.com/tendermint/go-common.PanicQ(0x5691c0, 0xc421314d40)
    /Users/smueller/development/projects/tendermint/code/src/github.com/tendermint/tendermint/vendor/github.com/tendermint/go-common/errors.go:44 +0xe0
github.com/tendermint/tendermint/consensus.(*ConsensusState).finalizeCommit(0xc42013e300, 0x17)
    /Users/smueller/development/projects/tendermint/code/src/github.com/tendermint/tendermint/consensus/state.go:1250 +0xe21
github.com/tendermint/tendermint/consensus.(*ConsensusState).tryFinalizeCommit(0xc42013e300, 0x17)
    /Users/smueller/development/projects/tendermint/code/src/github.com/tendermint/tendermint/consensus/state.go:1204 +0x134
github.com/tendermint/tendermint/consensus.(*ConsensusState).enterCommit.func1(0xc42013e300, 0x0, 0x17)
    /Users/smueller/development/projects/tendermint/code/src/github.com/tendermint/tendermint/consensus/state.go:1157 +0x64
github.com/tendermint/tendermint/consensus.(*ConsensusState).enterCommit(0xc42013e300, 0x17, 0x0)
    /Users/smueller/development/projects/tendermint/code/src/github.com/tendermint/tendermint/consensus/state.go:1184 +0x747
github.com/tendermint/tendermint/consensus.(*ConsensusState).addVote(0xc42013e300, 0x0, 0xc421209830, 0x0, 0x0, 0xc420151a20, 0x0, 0x0, 0xc420151aa0, 0x75e35, ...)
    /Users/smueller/development/projects/tendermint/code/src/github.com/tendermint/tendermint/consensus/state.go:1489 +0xb4c
github.com/tendermint/tendermint/consensus.(*ConsensusState).tryAddVote(0xc42013e300, 0x0, 0xc421209830, 0x0, 0x0, 0x0, 0x0)
    /Users/smueller/development/projects/tendermint/code/src/github.com/tendermint/tendermint/consensus/state.go:1386 +0x74
github.com/tendermint/tendermint/consensus.(*ConsensusState).handleMsg(0xc42013e300, 0x58c060, 0xc4212fc970, 0x0, 0x0, 0x17, 0x0, 0x6, 0xecff6d44d, 0x196560f0, ...)
    /Users/smueller/development/projects/tendermint/code/src/github.com/tendermint/tendermint/consensus/state.go:689 +0x66b
github.com/tendermint/tendermint/consensus.(*ConsensusState).receiveRoutine(0xc42013e300, 0x0)
    /Users/smueller/development/projects/tendermint/code/src/github.com/tendermint/tendermint/consensus/state.go:639 +0x43f
created by github.com/tendermint/tendermint/consensus.(*ConsensusState).OnStart
    /Users/smueller/development/projects/tendermint/code/src/github.com/tendermint/tendermint/consensus/state.go:327 +0x135
```

Testing against develop (0204d3c), yields a slightly different error:
```NOTE[12-29|14:17:52] Finalizing commit of block with 0 txs    module=consensus height=20 hash=965EF97EB3C41D134B2FDF1623DEAD415E96B27B root=
WARN[12-29|14:17:52] Error in proxyAppConn.BeginBlock         module=state error=EOF
panic: Paniced on a Sanity Check: updateToState() expected state height of 20 but found 19
goroutine 72 [running]:
panic(0x584220, 0xc42129a070)
    /usr/local/go/src/runtime/panic.go:500 +0x1a1
github.com/tendermint/tendermint/vendor/github.com/tendermint/go-common.PanicSanity(0x584220, 0xc42129a050)
    /Users/smueller/development/projects/tendermint/code/src/github.com/tendermint/tendermint/vendor/github.com/tendermint/go-common/errors.go:26 +0xe0
github.com/tendermint/tendermint/consensus.(*ConsensusState).updateToState(0xc420126680, 0xc42121ce40)
    /Users/smueller/development/projects/tendermint/code/src/github.com/tendermint/tendermint/consensus/state.go:524 +0xb66
github.com/tendermint/tendermint/consensus.(*ConsensusState).finalizeCommit(0xc420126680, 0x14)
    /Users/smueller/development/projects/tendermint/code/src/github.com/tendermint/tendermint/consensus/state.go:1296 +0xc5a
github.com/tendermint/tendermint/consensus.(*ConsensusState).tryFinalizeCommit(0xc420126680, 0x14)
    /Users/smueller/development/projects/tendermint/code/src/github.com/tendermint/tendermint/consensus/state.go:1224 +0x177
github.com/tendermint/tendermint/consensus.(*ConsensusState).enterCommit.func1(0xc420126680, 0x0, 0x14)
    /Users/smueller/development/projects/tendermint/code/src/github.com/tendermint/tendermint/consensus/state.go:1177 +0x64
github.com/tendermint/tendermint/consensus.(*ConsensusState).enterCommit(0xc420126680, 0x14, 0x0)
    /Users/smueller/development/projects/tendermint/code/src/github.com/tendermint/tendermint/consensus/state.go:1204 +0x746
github.com/tendermint/tendermint/consensus.(*ConsensusState).addVote(0xc420126680, 0xc421251e40, 0x0, 0x0, 0x1e0, 0x1da, 0x0)
    /Users/smueller/development/projects/tendermint/code/src/github.com/tendermint/tendermint/consensus/state.go:1484 +0x975
github.com/tendermint/tendermint/consensus.(*ConsensusState).tryAddVote(0xc420126680, 0xc421251e40, 0x0, 0x0, 0x20, 0x0)
    /Users/smueller/development/projects/tendermint/code/src/github.com/tendermint/tendermint/consensus/state.go:1382 +0x67
github.com/tendermint/tendermint/consensus.(*ConsensusState).handleMsg(0xc420126680, 0x5a8600, 0xc42013a8c0, 0x0, 0x0, 0x14, 0x0, 0x6, 0xecff70400, 0x8d6abf5, ...)
    /Users/smueller/development/projects/tendermint/code/src/github.com/tendermint/tendermint/consensus/state.go:726 +0x666
github.com/tendermint/tendermint/consensus.(*ConsensusState).receiveRoutine(0xc420126680, 0x0)
    /Users/smueller/development/projects/tendermint/code/src/github.com/tendermint/tendermint/consensus/state.go:676 +0x43f
created by github.com/tendermint/tendermint/consensus.(*ConsensusState).OnStart
    /Users/smueller/development/projects/tendermint/code/src/github.com/tendermint/tendermint/consensus/state.go:363 +0x48aNOTE[12-29|14:17:52] Finalizing commit of block with 0 txs    module=consensus height=20 hash=965EF97EB3C41D134B2FDF1623DEAD415E96B27B root=
WARN[12-29|14:17:52] Error in proxyAppConn.BeginBlock         module=state error=EOF
panic: Paniced on a Sanity Check: updateToState() expected state height of 20 but found 19
goroutine 72 [running]:
panic(0x584220, 0xc42129a070)
    /usr/local/go/src/runtime/panic.go:500 +0x1a1
github.com/tendermint/tendermint/vendor/github.com/tendermint/go-common.PanicSanity(0x584220, 0xc42129a050)
    /Users/smueller/development/projects/tendermint/code/src/github.com/tendermint/tendermint/vendor/github.com/tendermint/go-common/errors.go:26 +0xe0
github.com/tendermint/tendermint/consensus.(*ConsensusState).updateToState(0xc420126680, 0xc42121ce40)
    /Users/smueller/development/projects/tendermint/code/src/github.com/tendermint/tendermint/consensus/state.go:524 +0xb66
github.com/tendermint/tendermint/consensus.(*ConsensusState).finalizeCommit(0xc420126680, 0x14)
    /Users/smueller/development/projects/tendermint/code/src/github.com/tendermint/tendermint/consensus/state.go:1296 +0xc5a
github.com/tendermint/tendermint/consensus.(*ConsensusState).tryFinalizeCommit(0xc420126680, 0x14)
    /Users/smueller/development/projects/tendermint/code/src/github.com/tendermint/tendermint/consensus/state.go:1224 +0x177
github.com/tendermint/tendermint/consensus.(*ConsensusState).enterCommit.func1(0xc420126680, 0x0, 0x14)
    /Users/smueller/development/projects/tendermint/code/src/github.com/tendermint/tendermint/consensus/state.go:1177 +0x64
github.com/tendermint/tendermint/consensus.(*ConsensusState).enterCommit(0xc420126680, 0x14, 0x0)
    /Users/smueller/development/projects/tendermint/code/src/github.com/tendermint/tendermint/consensus/state.go:1204 +0x746
github.com/tendermint/tendermint/consensus.(*ConsensusState).addVote(0xc420126680, 0xc421251e40, 0x0, 0x0, 0x1e0, 0x1da, 0x0)
    /Users/smueller/development/projects/tendermint/code/src/github.com/tendermint/tendermint/consensus/state.go:1484 +0x975
github.com/tendermint/tendermint/consensus.(*ConsensusState).tryAddVote(0xc420126680, 0xc421251e40, 0x0, 0x0, 0x20, 0x0)
    /Users/smueller/development/projects/tendermint/code/src/github.com/tendermint/tendermint/consensus/state.go:1382 +0x67
github.com/tendermint/tendermint/consensus.(*ConsensusState).handleMsg(0xc420126680, 0x5a8600, 0xc42013a8c0, 0x0, 0x0, 0x14, 0x0, 0x6, 0xecff70400, 0x8d6abf5, ...)
    /Users/smueller/development/projects/tendermint/code/src/github.com/tendermint/tendermint/consensus/state.go:726 +0x666
github.com/tendermint/tendermint/consensus.(*ConsensusState).receiveRoutine(0xc420126680, 0x0)
    /Users/smueller/development/projects/tendermint/code/src/github.com/tendermint/tendermint/consensus/state.go:676 +0x43f
created by github.com/tendermint/tendermint/consensus.(*ConsensusState).OnStart
    /Users/smueller/development/projects/tendermint/code/src/github.com/tendermint/tendermint/consensus/state.go:363 +0x48a```

**Comment by srmo** [comment]
Edit: the develop error is usually preceded with the following log messages:
```
NOTE[12-29|14:30:11] Finalizing commit of block with 0 txs    module=consensus height=19 hash=E2A45AF971DDF2031AFF9BC54A1531D64D8F6EC7 root=
WARN[12-29|14:30:12] Stopping tmsp.socketClient for error: EOF module=tmspcli
WARN[12-29|14:30:12] Stopping tmsp.socketClient for error: EOF module=tmspcli
WARN[12-29|14:30:12] Stopping tmsp.socketClient for error: EOF module=tmspcli
NOTE[12-29|14:30:12] enterNewRound(20/0). Current: 20/0/RoundStepNewHeight module=consensus
NOTE[12-29|14:30:12] enterPrecommit: +2/3 prevoted proposal block. Locking module=consensus hash=41818B495FA19129EA6FDAA0396119651CDECBA9
NOTE[12-29|14:30:12] Finalizing commit of block with 0 txs    module=consensus height=20 hash=41818B495FA19129EA6FDAA0396119651CDECBA9 root=
WARN[12-29|14:30:12] Error in proxyAppConn.BeginBlock         module=state error=EOF
panic: Paniced on a Sanity Check: updateToState() expected state height of 20 but found 19
```

**Comment by akalyaev** [comment]
I believe this is the place - https://github.com/tendermint/tendermint/blob/b74a97a4f63f1627c907528ff00b7e1d6175f31f/consensus/state.go#L1279 Before we used to call `PanicQ` here (in `master`).

There is a [note saying that we should wait for an app to come back](https://github.com/tendermint/tendermint/blob/b74a97a4f63f1627c907528ff00b7e1d6175f31f/state/execution.go#L41) in case of error, but I don't see how we're doing it (if we do)?

**Comment by akalyaev** [comment]
"This is expected. At some point we need to implement allowing tendermint to sit pretty while the app restarts. Easier said than done but we'll aim to get it into 0.9"

**Comment by ebuchman** [comment]
This is fixed on develop. Soon to be released as 0.9. Now when the app dies Tendermint logs an error and politely asks to be restarted. In the future, we'll make it possible to recover without tendermint restarting. Closing.
